pep-0675.rst

```diff
@@ -0,0 +1,954 @@
+
+    # Fetch all users (since 1 = 1 is always true).
+    query_user(conn, "user123 OR 1 = 1")
+
+
+To prevent such SQL injection attacks, SQL APIs offer parameterized
+queries, which separate the executed query from user-controlled data
+and make it impossible to run arbitrary queries. For example, with
+`sqlite3 <https://docs.python.org/3/library/sqlite3.html>`_, our
+original function would be written safely as a query with parameters:
+
+::
+
+    def query_user(conn: Connection, user_id: str) -> User:
+        query = "SELECT * FROM data WHERE user_id = ?"
+        conn.execute(query, (user_id,))
+
+
+The problem is that there is no way to enforce this
+discipline. sqlite3's own `documentation
+<https://docs.python.org/3/library/sqlite3.html>`_ can only admonish
+the reader to not dynamically build the ``sql`` argument from external
+input; the API's authors cannot express that through the type
+system. Users can (and often do) still use a convenient f-string as
+before and leave their code vulnerable to SQL injection.
+
+Existing tools, such as the popular security linter `Bandit
+<https://github.com/PyCQA/bandit/blob/aac3f16f45648a7756727286ba8f8f0cf5e7d408/bandit/plugins/django_sql_injection.py#L102>`_,
+attempt to detect unsafe external data used in SQL APIs, by inspecting
+the AST or by other semantic pattern-matching. These tools, however,
+preclude common idioms like storing a large multi-line query in a
+variable before executing it, adding literal string modifiers to the
+query based on some conditions, or transforming the query string using
+a function. (We survey existing tools in the "Rejected Alternatives"
+section.) For example, many tools will detect a false positive issue
+in this benign snippet:
+
+
+::
+
+    def query_data(conn: Connection, user_id: str, limit: bool) -> None:
+        query = """
+            SELECT
+                user.name,
+                user.age
+            FROM data
+            WHERE user_id = ?
+        """
+        if limit:
+            query += " LIMIT 1"
+
+        conn.execute(query, (user_id,))
+
+We want to forbid harmful execution of user-controlled data while
+still allowing benign idioms like the above and not requiring extra
+user work.
+
+To meet this goal, we introduce the ``Literal[str]`` type, which only
+accepts string values that are known to be made of literals. This is a
+generalization of the ``Literal["foo"]`` type from `PEP 586
+<https://www.python.org/dev/peps/pep-0586/>`_. A string of type
+``Literal[str]`` cannot contain user-controlled data. Thus, any API
+that only accepts ``Literal[str]`` will be immune to injection
+vulnerabilities (with pragmatic `limitations <Appendix B:
+Limitations_>`_).
+
+Since we want the ``sqlite3`` ``execute`` method to disallow strings
+built with user input, we make it accept a ``sql`` query that is of
+type ``Literal[str]``:
+
+::
+
+    def execute(self, sql: Literal[str], parameters: Iterable[str] = ...) -> Cursor: ...
```

Comment:
Where would you make this change? Would typeshed's stubs for sqlite change?

Reply:
Yes, we would have to update its typeshed stub. Linked to it.